lib/iris/analysis/_grid_angles.py

```diff
@@ -0,0 +1,390 @@
+
+        # TEMPORARY CHECKS:
+        # ensure that the two unit vectors are perpendicular.
+        dotprod = np.sum(phi_hatvec * lmb_hatvec, axis=0)
+        assert np.allclose(dotprod, 0.0)
+        # ensure that the vector components carry the original magnitude.
+        mag_orig = np.sum(pr * pr)
+        mag_rot = np.sum(pr_cmpt_e * pr_cmpt_e) + np.sum(pr_cmpt_n * pr_cmpt_n)
+        rtol = 1.e-3
+        check = np.allclose(mag_rot, mag_orig, rtol=rtol)
+        if not check:
+            print (mag_rot, mag_orig)
+            assert np.allclose(mag_rot, mag_orig, rtol=rtol)
+
+    return psi
+
+
+def gridcell_angles(x, y=None, cell_angle_boundpoints='mid-lhs, mid-rhs'):
+    """
+    Calculate gridcell orientations for an arbitrary 2-dimensional grid.
+
+    The input grid is defined by two 2-dimensional coordinate arrays with the
+    same dimensions (ny, nx), specifying the geolocations of a 2D mesh.
+
+    Input values may be coordinate points (ny, nx) or bounds (ny, nx, 4).
+    However, if points, the edges in the X direction are assumed to be
+    connected by wraparound.
+
+    Input can be either two arrays, two coordinates, or a single cube
+    containing two suitable coordinates identified with the 'x' and'y' axes.
+
+    Args:
+
+    The inputs (x [,y]) can be any of the folliwing :
+
+    * x (:class:`~iris.cube.Cube`):
+        a grid cube with 2D X and Y coordinates, identified by 'axis'.
+        The coordinates must be 2-dimensional with the same shape.
+        The two dimensions represent grid dimensions in the order Y, then X.
+
+    * x, y (:class:`~iris.coords.Coord`):
+        X and Y coordinates, specifying grid locations on the globe.
+        The coordinates must be 2-dimensional with the same shape.
+        The two dimensions represent grid dimensions in the order Y, then X.
+        If there is no coordinate system, they are assumed to be true
+        longitudes and latitudes.  Units must convertible to 'degrees'.
+
+    * x, y (2-dimensional arrays of same shape (ny, nx)):
+        longitude and latitude cell center locations, in degrees.
+        The two dimensions represent grid dimensions in the order Y, then X.
+
+    * x, y (3-dimensional arrays of same shape (ny, nx, 4)):
+        longitude and latitude cell bounds, in degrees.
+        The first two dimensions are grid dimensions in the order Y, then X.
+        The last index maps cell corners anticlockwise from bottom-left.
+
+    Optional Args:
+
+    * cell_angle_boundpoints (string):
+        Controls which gridcell bounds locations are used to calculate angles,
+        if the inputs are bounds or bounded coordinates.
+        Valid values are 'lower-left, lower-right', which takes the angle from
+        the lower left to the lower right corner, and 'mid-lhs, mid-rhs' which
+        takes an angles between the average of the left-hand and right-hand
+        pairs of corners.  The default is 'mid-lhs, mid-rhs'.
```

Comment:
The original code I was working from (ex-IDL) used only point values.
I extended it to use bounds because that is much more logical, and also works properly for _regional_ data.
What I found was that, while 'mid-lhs..mid-rhs' was a more obvious choice, 'bottom-left..bottom-right' was closer to the 'official' cell angles in the model output file (when included).
So I hedged my bets by making it switchable, but only supported 2 options = 'seems most logical' and 'best match to official values'.

We can perhaps standardise on just the 'mid-lhs..mid-rhs' option, if I could convince myself that the differences are not significant.  I don't really want to make 'bottom-left..bottom-right' the default, as it seems the poorer choice.

Reply:
>  standardise on just the 'mid-lhs..mid-rhs' option, if I could convince myself that the differences are not significant

But unfortunately I checked, and the discrepancies *are* significant in some areas.  At least a few degrees for practical purposes.
Also the errors for quasi-ideal square-ish gridcells rotated at various angles seem to be ~linear in the size of the cells for the 'bottom-left..bottom-right' method while ~square for the 'mid-lhs..mid-rhs'.
So that means I'm not comfortable 'standardising' on either of these exclusively :disappointed: